Documentation/designs/agent-deployment-pattern-sidecar.md

```diff
@@ -0,0 +1,116 @@
+
+## Summary
+
+With Prometheus-Operator finally supporting running Prometheus in Agent mode, we can start thinking about different deployment patterns that can be explored with this minimal container. This document aims to continue the work started by [this document](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/designs/prometheus-agent.md), focusing on exploring how Prometheus-Operator can leverage deploying PrometheusAgents as sidecars running alongside pods that a user wants to monitor.
+
+## Background
+
+By the time this document was written, Prometheus-Operator can deploy Prometheus in Agent mode, but only using a pattern similar to the original implementation of Prometheus Server: using StatefulSets. The [original design document](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/designs/prometheus-agent.md) for Prometheus Agent already mentions that different deployment patterns are desired, however, for the sake of speeding up the initial implementation it was decided to re-use the logic and start with the Agent running as StatefulSets.
+
+Also for the sake of speeding up implementation, this document won't focus on several new Deployment patterns, but only one: Sidecar Injection.
+
+Looking at the traditional deployment model, we have a single Prometheus (or an HA setup) per cluster or namespace, responsible for scraping all containers under their scope. Prometheus operator relies on `ServiceMonitor`, `PodMonitor`, and `Probe` CRs to configure Prometheus, which will eventually use Kubernetes service-discovery to find endpoints that need to be scraped.
+
+Depending on the Cluster's scale and how often Prometheus hits Kubernetes API, Prometheus service discovery can increase the load on the API significantly and affect the overall functionality of said cluster.
+
+Another problem is that one or more containers can be updated to a problematic version that causes a [Cardinality Spike](https://grafana.com/blog/2022/02/15/what-are-cardinality-spikes-and-why-do-they-matter/). Depending on the proportion of the spike, it is possible that a container could single-handedly crash the monitoring system of the whole cluster.
+
+![Traditional Deployment Pattern](assets/agent-deployment-pattern-sidecar/traditional-deployment-pattern.png "Traditional Deployment Pattern").
+
+## Proposal
+
+This document proposes a new deployment model where Prometheus-Operator injects Prometheus agents as a sidecar container (and Prometheus config reloader) to pods that needs to be scrapped. With a sidecar, we tackle both problems mentioned above:
+
+> * Load on Kubernetes API won't exist since it's not needed anymore. Prometheus will scrape containers from the same pod through their shared network interface and scrape configuration can be declared via pod annotations.
+> * A sudden cardinality spike will not affect the whole monitoring system. In a worst-case scenario, it will fail a single pod.
```

Comment:
I think this isn't entirely true. While an agent monitoring only a single pod is less likely to buckle under a cardinality increase, by removing this bottleneck we expose the next layer up to the same increase. That may or may not have a higher overall capacity, but we massively increased the capacity of the layer feeding into it, so we may be exposing it to _more_ cardinality issues.

To illustrate, consider a multi cluster setup with an agent setup per cluster, feeding into a global observability layer. Previously, deploying a cardinality explosion to one cluster might overload the agent there, signalling that there is a problem with this deploy. Now, we create the capacity for this cluster alone to blow up global observability.

How exactly this plays out really depends on the details, so I don't think this model would be _worse_. It opens up a different set of tradeoffs which is generally helpful.

Reply:
I see your point, but that was not what I had in mind in this statement 😅 

The goal of the sidecar injection is to scale horizontally, making sure that a cardinality increase that is capable of OOMing Prometheus will prevent us from retrieving metrics from a single pod, but not from the entire cluster.